### PR TITLE
Alternative side-effects solution for async migrations

### PR DESCRIPTION
**Pull request by tiina303**
## Changes

For better clarity and more flexibility in the future changing the side-effect concept to be AsyncMigrationOperation taking function pointers instead.

## How did you test this code?

CI

### CODE REVIEW

**Comment by fuziontech** [comment]
I love the idea of passing `query_id` or some context into either the side effect function or have the runner have this context and log the context with any errors. That should definitely be added.

**Review by yakkomajuri — COMMENTED** [review]
I think this is valid, but I'm not fond of the implementation. I'd like these migrations to be clean and easy to read, with more advanced options available for those that would like more granular control.

Thus, let's:

1. Remove `side_effect` stuff
2. Rename `fn` to something better maybe?
3. Keep all the other stuff, checking that each operation only has one of `sql` or `fn` 

Then we can keep everything as is, yet people needing granular control write their own function with their own e.g. `sync_execute` calls and all with more flexibility. I also like the idea of passing it some params.

**Comment by tiina303** [comment]
>  I'd like these migrations to be clean and easy to read, with more advanced options available for those that would like more granular control.

Good point. I'll add a constructor that's easier to read and similar to the way we currently define ops, which will just do the boilerplate stuff for single op/rollback.

File: posthog/async_migrations/runner.py [hunk 1]
```diff
@@ -173,15 +170,9 @@ def attempt_migration_rollback(migration_instance: AsyncMigration, force: bool =
     try:
         ops = get_async_migration_definition(migration_instance.name).operations
 
-        i = len(ops) - 1
-        for op in reversed(ops[: migration_instance.current_operation_index + 1]):
-            if not op.rollback:
-                if op.rollback == "":
-                    continue
-                raise Exception(f"No rollback provided for operation at index {i}: {op.sql}")
```
Comment (tiina303) [inline posthog/async_migrations/runner.py:181]: Note that sometimes the rollback is do nothing & that's the new default. All operations should have a rollback, if they don't however given that we can now provide random functions the `rollback_fn` could raise an Exception.

File: posthog/async_migrations/runner.py [hunk 2]
```diff
@@ -173,15 +170,9 @@ def attempt_migration_rollback(migration_instance: AsyncMigration, force: bool =
...
+        for op_index in range(migration_instance.current_operation_index, -1, -1):
+            op = ops[op_index]
             execute_op(op, str(UUIDT()), rollback=True)
-            op.side_effect_rollback()
```
Comment (tiina303) [inline posthog/async_migrations/runner.py:183]: Note that we also did this in a bad location before (this pr fixes that). E.g. 0002 last step would detach table before turning compute materialized columns off but we should do that in reverse order.

**Comment by yakkomajuri** [comment]
First thought: for simplicity, can't we let the constructor handle if it is a simple op or not?

**Comment by tiina303** [comment]
> First thought: for simplicity, can't we let the constructor handle if it is a simple op or not?

python doesn't let us have multiple constructors /function overload, so options are:
1. classmethod (factory method as they are called)
2. subclass with a constructor and nothing else
3. have args for everything in one constructor - I find that more confusing/harder to use generally

File: posthog/async_migrations/runner.py (hunk 1 above)
Comment (yakkomajuri) [reply posthog/async_migrations/runner.py:181]: That's why this error was here. To force people to write a rollback or rather override it explicitly to a no-op

**Review by yakkomajuri — APPROVED** [review]
left a comment about the rollback and otherwise I don't like the naming of things in this PR.

`fn`, `simple_op` and stuff. But I can't come up with anything better right now so lgtm

**Comment by yakkomajuri** [comment]
Also, have you tested running a migration through manually after the changes? Would be good to not rely only on tests

**Comment by tiina303** [comment]
> Also, have you tested running a migration through manually after the changes? Would be good to not rely only on tests

nope 😅 

> `fn`, `simple_op` and stuff. But I can't come up with anything better right now so lgtm

one of the 2 hard things in computer science 😅  open to suggestions later too if you think of anything.

> That's why this error was here. To force people to write a rollback or rather override it explicitly to a no-op

If we want to error on that, let's do that on our side. i.e. one can't write a migration without all steps being able to roll back.

**Pull request merged**